bcda/auth/oktabackend.go

```diff
@@ -0,0 +1,321 @@
+package auth
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var ob *OB
+var publicKeys map[string]rsa.PublicKey
+var once sync.Once
+
+var apiBase string
+var authString string
+var authServerID string
+
+func init() {
+	log.SetFormatter(&log.JSONFormatter{})
+	SetProvider(strings.ToLower(os.Getenv(`BCDA_AUTH_PROVIDER`)))
+}
+
+type OB struct{}
+
+func NewOB() *OB {
+	once.Do(func() {
+		ob = &OB{}
+		publicKeys = getPublicKeys()
+		apiBase = os.Getenv("OKTA_CLIENT_ORGURL")
+		authServerID = os.Getenv("OKTA_OAUTH_SERVER_ID")
+		authString = fmt.Sprintf("SSWS %s", os.Getenv("OKTA_CLIENT_TOKEN"))
+		go refreshKeys()
+	})
+	return ob
+}
+
+func (ob *OB) PublicKeyFor(id string) (rsa.PublicKey, bool) {
+	key, ok := publicKeys[id]
+	log.Printf("invalid key id %s presented", id)
+	return key, ok
+}
+
+func (ob *OB) AddClientApplication(localID string) (string, string, error) {
```

Comment:
What do you think about moving this func to `oktaclient.go`?  That way, we could consolidate the external communication with okta in one place: the logging of okta requests/responses, the handling of errors, the building of requests and headers, etc.

Reply:
I'm thinking single responsibility principle here